torch/fx/experimental/efficient_conv_bn_eval.py

```diff
@@ -0,0 +1,143 @@
+        return output
+    else:
+        conv_out = conv._conv_forward(x, conv.weight, conv.bias)
+        return bn(conv_out)
+
+
+def efficient_conv_bn_eval_graph_transform(fx_model):
+    """Find consecutive conv+bn calls in the graph, inplace modify the graph
+    with the fused operation."""
+    modules = dict(fx_model.named_modules())
+
+    patterns = [(torch.nn.modules.conv._ConvNd, torch.nn.modules.batchnorm._BatchNorm)]
+
+    pairs = []
+    # Iterate through nodes in the graph to find ConvBN blocks
+    for node in fx_model.graph.nodes:
+        # If our current node isn't calling a Module then we can ignore it.
+        if node.op != "call_module":
+            continue
+        target_module = modules[node.target]
+        found_pair = False
+        for conv_class, bn_class in patterns:
+            if isinstance(target_module, bn_class):
+                target = node.args[0].target
+                if target not in modules:
+                    continue
+                source_module = modules[target]
+                if isinstance(source_module, conv_class):
+                    found_pair = True
+        # Not a conv-BN pattern or output of conv is used by other nodes
+        if not found_pair or len(node.args[0].users) > 1:
+            continue
+
+        # Find a pair of conv and bn computation nodes to optimize.
+        # To avoid modifying the graph during traversal, we record all pairs
+        # first, and then optimize them later.
+        conv_node = node.args[0]
+        bn_node = node
+        if modules[bn_node.target].track_running_stats:
+            pairs.append([conv_node, bn_node])
+
+    for conv_node, bn_node in pairs:
+        # set insertion point
+        fx_model.graph.inserting_before(conv_node)
+        # create `get_attr` node to access modules
+        # note that we directly call `create_node` to fill the `name`
+        # argument. `fx_model.graph.get_attr` and
+        # `fx_model.graph.call_function` does not allow the `name` argument.
+        conv_get_node = fx_model.graph.create_node(
+            op="get_attr", target=conv_node.target, name="get_conv"
+        )
+        bn_get_node = fx_model.graph.create_node(
+            op="get_attr", target=bn_node.target, name="get_bn"
+        )
+        # prepare args for the fused function
+        args = (bn_get_node, conv_get_node, conv_node.args[0])
+        # create a new node
+        new_node = fx_model.graph.create_node(
+            op="call_function",
+            target=efficient_conv_bn_eval_control,
+            args=args,
+            name="efficient_conv_bn_eval",
+        )
+        # this node replaces the original conv + bn, and therefore
+        # should replace the uses of bn_node
+        bn_node.replace_all_uses_with(new_node)
+        # take care of the deletion order:
+        # delete bn_node first, and then conv_node
+        fx_model.graph.erase_node(bn_node)
+        fx_model.graph.erase_node(conv_node)
```

Comment:
I believe this is unsound because we aren't checking `len(conv_node.users)==1`.  It is also possible the input gets mutated in-place between the ops.  

Note the pattern_matcher_pass stuff in inductor handles this for you.

Reply:
The check of `len(conv_node.users)==1` is [here](https://github.com/pytorch/pytorch/pull/108757/files#diff-43ff9859f287716c8447d1cd78b86ccd6b590cc5ed53609ce9a93092b4142eb9R100). The inplace mutation is indeed a good point. I will use `pattern_matcher_pass` to ensure the soundness.